[0.9.3 built off master] Inserting value in scientific notation with a trailing i causes panic

Inserting a point with value in scientific notation with a trailing `i` causes the db to panic.

In the CLI

```
> insert sci value=9e10i
ERR: Post http://localhost:8086/write?consistency=any&db=mydb&precision=n&rp=: EOF
```

causes panic

```
panic: unable to parse number value '9e10i': strconv.ParseInt: parsing "9e10": invalid syntax

goroutine 56 [running]:
github.com/influxdb/influxdb/tsdb.newFieldsFromBinary(0xc20805a405, 0xb, 0x1fb, 0x8)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:1061 +0x63d
github.com/influxdb/influxdb/tsdb.(*point).unmarshalBinary(0xc20822c080, 0x8)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:949 +0x3b
github.com/influxdb/influxdb/tsdb.(*point).Fields(0xc20822c080, 0xc20810d9b0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:896 +0x28
github.com/influxdb/influxdb/tsdb.(*Shard).validateSeriesAndFields(0xc2080b4880, 0xc2085663c0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/shard.go:337 +0x5f7
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc2080b4880, 0xc2085663c0, 0x1, 0x1, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/shard.go:152 +0x75
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc20809a780, 0x266, 0xc2085663c0, 0x1, 0x1, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/store.go:281 +0xf7
github.com/influxdb/influxdb/cluster.func·002(0x266, 0x1, 0xc2085663c0, 0x1, 0x1)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:241 +0xfc
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:271 +0x30b
```

Similarly

```
insert sci value=9ie10
```

also causes panic

```
panic: unable to parse number value '9ie10': strconv.ParseFloat: parsing "9ie10": invalid syntax

goroutine 55 [running]:
github.com/influxdb/influxdb/tsdb.newFieldsFromBinary(0xc208052405, 0xb, 0x1fb, 0x8)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:1061 +0x63d
github.com/influxdb/influxdb/tsdb.(*point).unmarshalBinary(0xc2080ac400, 0x8)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:949 +0x3b
github.com/influxdb/influxdb/tsdb.(*point).Fields(0xc2080ac400, 0xc2080c19b0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/points.go:896 +0x28
github.com/influxdb/influxdb/tsdb.(*Shard).validateSeriesAndFields(0xc2080ac500, 0xc208398d30, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/shard.go:337 +0x5f7
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc2080ac500, 0xc208398d30, 0x1, 0x1, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/shard.go:152 +0x75
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc208092780, 0x266, 0xc208398d30, 0x1, 0x1, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/tsdb/store.go:281 +0xf7
github.com/influxdb/influxdb/cluster.func·002(0x266, 0x1, 0xc208398d30, 0x1, 0x1)
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:241 +0xfc
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
	/Users/michaeldesa/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:271 +0x30b
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -28,6 +28,7 @@
 - [#3530](https://github.com/influxdb/influxdb/pull/3530): Aliasing a column no longer works
 - [#3436](https://github.com/influxdb/influxdb/issues/3436): Fix panic in hinted handoff queue processor
 - [#3401](https://github.com/influxdb/influxdb/issues/3401): Derivative on non-numeric fields panics db
+- [#3583](https://github.com/influxdb/influxdb/issues/3583): Inserting value in scientific notation with a trailing i causes panic
 
 ## v0.9.2 [2015-07-24]
 

--- a/tsdb/points.go
+++ b/tsdb/points.go
@@ -587,7 +587,7 @@ func scanNumber(buf []byte, i int) (int, error) {
 		}
 		i += 1
 	}
-	if isInt && decimals > 0 {
+	if isInt && (decimals > 0 || scientific) {
 		return i, fmt.Errorf("invalid number")
 	}
 
@@ -596,6 +596,10 @@ func scanNumber(buf []byte, i int) (int, error) {
 	// if we should parse the number to the actual type.  It does not do it all the time because it incurs
 	// extra allocations and we end up converting the type again when writing points to disk.
 	if isInt {
+		// Make sure the last char is an 'i' for integers (e.g. 9i10 is not valid)
+		if buf[i-1] != 'i' {
+			return i, fmt.Errorf("invalid number")
+		}
 		// Parse the int to check bounds the number of digits could be larger than the max range
 		// We subtract 1 from the index to remove the `i` from our tests
 		if len(buf[start:i-1]) >= maxInt64Digits || len(buf[start:i-1]) >= minInt64Digits {

--- a/tsdb/points_test.go
+++ b/tsdb/points_test.go
@@ -486,6 +486,19 @@ func TestParsePointBooleanInvalid(t *testing.T) {
 	}
 }
 
+func TestParsePointScientificIntInvalid(t *testing.T) {
+	_, err := tsdb.ParsePointsString(`cpu,host=serverA,region=us-west value=9ie10`)
+	if err == nil {
+		t.Errorf(`ParsePoints("%s") mismatch. got nil, exp error`, `cpu,host=serverA,region=us-west value=9ie10`)
+	}
+
+	_, err = tsdb.ParsePointsString(`cpu,host=serverA,region=us-west value=9e10i`)
+	if err == nil {
+		t.Errorf(`ParsePoints("%s") mismatch. got nil, exp error`, `cpu,host=serverA,region=us-west value=9e10i`)
+	}
+
+}
+
 func TestParsePointUnescape(t *testing.T) {
 	test(t, `foo\,bar value=1i`,
 		tsdb.NewPoint(

